lec08-stability.Rmd

```diff
@@ -1,4 +1,620 @@
+  # Calculates dN/dt for the logistic equation
+  
+  # t: time point at which to evaluate derivative
+  # state: vector of variables (here it's just N)
+  # parameters: vector of model parameters c(r, K)
+  
+  N <- state 
+  
+  r <- parameters['r'] # get the element labelled 'r'
+  K <- parameters['K'] # get the element labelled 'K'
+
+  #rate of change
+  dN <- r * N * (1 - N / K)
+    
+  #return rate of change
+  return(list(c(dN)))
+}
+```
+
+
+```{r}
+# Run ode
+
+library(dplyr)
+library(tidyr)
+library(deSolve)
+
+# define parameters for the ode function
+parameters <- c(r = 0.5, K = 50)
+initial_conditions <- seq(10, 80, by = 10)
+times <- seq(0, 15, by = 0.01)
+
+# run ode inside a loop to calculate trajectories for several initial conditions
+results <- data_frame(initial_conditions = seq(10, 80, by = 10)) %>% # CTRL-shift-m to make pipe
+  # 'do' is a dplyr function that returns a dataframe after doing a series of computations.
+  do(data.frame( # 'data.frame' converts the output of 'ode' to a dataframe
+    ode(
+      y = initial_conditions, # each time the 'do' loop runs, a new initial condition will be used
+      times = times,
+      func = logistic_fn,
+      parms = parameters
+      )
+  ))
+
+# display the first few rows of the results
+head(results)  
+```
+
+This is exactly what we wanted. We have chosen a set of initial conditions 
+(`seq(10, 80, by = 10)`) and created a dataframe that has one column of times (`time`)
+and a column for each trajectory resulting from each initial condition (`X1`, `X2`, ...). 
+Now we can use `ggplot` to plot all the trajectories.
+
+```{r}
+# make a plot
+library(ggplot2)
+
+results %>%
+    gather(Column, Value, -time) %>%
+    ggplot(aes(x = time, y = Value, color = Column)) +
+    geom_line(aes(x = time, y = Value)) +
+    labs(y = "Population size")
+```
+
+This plot very nicely summarizes the behaviour of the logistic equation: it shows the
+path that the population size $N$ will take in time from several initial conditions. 
+Importantly, the carrying capacity $K$ which we set to be $50$ is clearly a stable 
+fixed point: all the trajectories go towards $K$. 
+
+#### Challenge
+
+Modify the plotting code above so that the legend lists the initial conditions for each
+line instead of `X1`, `X2`, etc. 
+
+```{r, eval=FALSE, echo=FALSE}
+# Challenge solution
+
+results %>%
+    gather(Column, Value, -time) %>%
+    mutate(Column = gsub("X", "", Column) %>% # add this
+    gsub("$", "0", .)) %>% # and this
+    ggplot(aes(x = time, y = Value, color = Column)) +
+    geom_line(aes(x = time, y = Value)) +
+    labs(y = "Population size")
+
+# or
+
+results %>%
+    rename(Condition10 = X1, Condition20 = X2) %>% # etc
+    gather(Column, Value, -time) %>%
+    ggplot(aes(x = time, y = Value, color = Column)) +
+    geom_line(aes(x = time, y = Value)) +
+    labs(y = "Population size")
+
+# or add this line before the plot
+
+names(results) <- c('time', data.matrix(results[1, -1]))
```

Comment:
Note that this modifies the columns of the `results` data frame, while the other two only touches the modified data frame that is being plotted. It doesn't matter in this lecture since you don't use `results` again, but otherwise it could be worse/better depending on if you want those columns to have new names or not =p

Reply:
Does this actually work? It's a bit unclear based on the functions. Never heard of or used `data.matrix` before.

This is a better way of doing it.

```
names(results) <- c("time", "Prey", "Predator")
# Or
results <- setNames(results, c("time", "Prey", "Predator"))
```

Reply:
I just use `data.matrix` to get rid of the column names and only return the values of the first row (which are the initial conditions). You could also use `unlist(row.names=FALSE`, which is more explicit, but also more typing =) So I set the column names to '10', '20', '30', etc., not 'Prey' and 'Predator'.

Reply:
Try it, it works!

Reply:
Oh, deerrrrr, I was thinking it was a different code set... I should have looked closer.